MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,247 @@
+appears in multiple pack-files covered by the multi-pack-index, we select the
+pack-file with the most-recent _modified time_ as reported by the filesystem.
+For example, if someone downloads a second large prefetch pack when prefetching
+for the first time in 31 days, then that large prefetch pack is referenced and
+all older prefetch packs are "stale" and not referenced by the multi-pack-index.
+
+We can extend the `git multi-pack-index` builtin to delete pack-files that have
+no objects referenced by the multi-pack-index. This should happen at the same
+time as we update the multi-pack-index since we don't want to have this list of
+pack-files in the multi-pack-index if they don't exist. We also don't want to
+remove them from the multi-pack-index while the pack-files still exist, because
+then a Git process will start referencing those pack-files in the `packed_git`
+linked list.
+
+### Repack the prefetch pack-files
+
+We can reduce the number of prefetch packs and improve file locality by
+combining prefetch packs into new "repacked" pack-files. Use `git show-index
+<object-dir>/pack/prefetch-<timestamp[i]>-<hash[i]>.idx` to list the objects
+in the prefetch packs and pipe that output to `git pack-objects -q
+<object-dir>/pack/repacked-<timestamp>`.
+
+We should refrain from packing the most-recent prefetch packs, as we don't want
+to get in the way of the prefetch machinery for determining the most-recent
+timestamp. We could, however, start repacking the prefetch packs from
+oldest-to-newest, stopping in the most-recent week, or when the prefetch packs
+are large enough to satisfy a batch. Some stopping conditions could include
+"pack-file size is above 1GB".
+
+Scheduling Actions
+------------------
+
+We already have a prefetch action that is automatically queued in the
+background. In addition, a `gvfs prefetch --commits` action sends a message
+to the mount process to run a post-fetch job (compute multi-pack-index and
+commit-graph). These actions already have some logic that we will want to
+replicate, so we should make them explicit requirements of an action:
+
+* Most actions intend to have some delay between successful runs. Use a
+  `IsReady(DateTime time)` method to perform only the check if the given
+  `time` is long enough after the previous run. (This is most likely stored
+  as a timestamp in a list of pack-files or as the content of a file in
+  the shared object directory.)
+* Most logic happens during an `Execute()` method.
+* If the mount process needs to shut down while the action is active, we need
+  a `Stop()` method for the action to interrupt itself. This includes halting
+  any Git processes that are being run.
+
+The actions are then run in sequence by an `ActionRunner` that triggers a run
+even on some interval, runs the actions that are ready, and tracks the
+currently-running action (so the mount process can tell the `ActionRunner` to

```

Comment:
Please make sure the new ActionRunner checks if the device is ready before running to address #481

Reply:
Thanks for that. I just pushed to #493 a commit that includes a check for the device being ready. It's a bit crude, but perhaps it is enough? [Please let me know what you think](https://github.com/Microsoft/VFSForGit/pull/493/commits/8226cb77bfe99ba206b4b6a3a59a3be6e11d064c).